contracts/mapping-token/contracts/v2/Backing.sol

```diff
@@ -0,0 +1,268 @@
+    // [24, 32) bytes: Nonce
+    function encodeMessageId(uint32 bridgedLanePosition, uint64 nonce) public view returns (uint256) {
+        return uint256(remoteChainPosition) << 96 + uint256(bridgedLanePosition) << 64 + uint256(nonce);
+    }
+
+    // here add InBoundLane, and remote issuing module must add the corresponding OutBoundLane
+    function addInboundLane(address mappingTokenFactory, address inboundLane) external onlyOwner {
+        uint32 bridgedChainPosition = IMessageVerifier(inboundLane).bridgedChainPosition();
+        require(remoteChainPosition == bridgedChainPosition, "Invalid bridged chain position");
+        uint32 bridgedLanePosition = IMessageVerifier(inboundLane).bridgedLanePosition();
+        inboundLanes[bridgedLanePosition] = inboundLane;
+        emit NewInBoundLaneAdded(mappingTokenFactory, inboundLane);
+    }
+
+    // here add OutBoundLane, and remote issuing module must add the corresponding InBoundLane
+    function addOutboundLane(address outboundLane) external onlyOwner {
+        uint32 bridgedChainPosition = IMessageVerifier(outboundLane).bridgedChainPosition();
+        require(remoteChainPosition == bridgedChainPosition, "Invalid bridged chain position");
+        uint32 bridgedLanePosition = IMessageVerifier(outboundLane).bridgedLanePosition();
+        outboundLanes[bridgedLanePosition] = outboundLane;
+        emit NewOutBoundLaneAdded(bridgedLanePosition, outboundLane);
+    }
+
+    /**
+     * @notice reigister new erc20 token to the bridge. Only owner can do this.
+     * @param bridgedLanePosition the bridged lane positon, this register message will be delived to this lane position
+     * @param token the original token address
+     * @param name the name of the original token
+     * @param symbol the symbol of the original token
+     * @param decimals the decimals of the original token
+     */
+    function registerErc20Token(
+        uint32 bridgedLanePosition,
+        address token,
+        string memory name,
+        string memory symbol,
+        uint8 decimals
+    ) external payable onlyOwner {
+        require(registeredTokens[token] == false, "Backing: token has been registered");
+
+        address outboundLane = outboundLanes[bridgedLanePosition];
+        require(outboundLane != address(0), "cannot find outboundLane to send message");
+        bytes memory newErc20Contract = abi.encodeWithSelector(
+            IMappingTokenFactory.newErc20Contract.selector,
+            address(this),
+            ERC20_TOKEN_TYPE,
+            token,
+            thisChainName,
+            name,
+            symbol,
+            decimals
+        );
+        uint256 fee = IFeeMarket(feeMarket).market_fee();
+        require(msg.value >= fee, "not enough fee to pay");
+        uint64 nonce = IOutboundLane(outboundLane).send_message{value: fee}(remoteMappingTokenFactory, newErc20Contract);
+        uint256 messageId = encodeMessageId(bridgedLanePosition, nonce);
```

Comment:
Message layer also define a message key, Can we use the same key uniformly?

Reply:
Yes we can use the same key

Reply:
OK, I will change the returns value to message key.

Reply:
I will use the encode func from contract MessageVerifier.

Reply:
I mean, I will return the encoded message key. And the app layer doesn't have to deal with the encode thing.

Reply:
nonce => messageKey ?

Reply:
Yes!

Reply:
OK.